### PR TITLE
Add TF port of BLIP

### PR DESCRIPTION
**Pull request by Rocketknight1**
Work in progress right now, will update this when it's closer to being ready!

### CODE REVIEW

**Comment by Rocketknight1** [comment]
The TF port is mostly complete now and tests are passing locally - I just need to go around updating docs and auto classes and so on. The main code should be ready for review!

File: src/transformers/models/blip/modeling_tf_blip.py [hunk 1]
```diff
@@ -0,0 +1,1778 @@
...
+        r"""
+        Overrides *generate* function to be able to use the model as a conditional generator
+
+        Parameters:
+            pixel_values (*tf.Tensor* of shape *(batch_size, image_width, image_height)*:
+                Input image to be processed
+            input_ids (*tf.Tensor* of shape *(batch_size, sequence_length)*, *optional*):
+                The sequence used as a prompt for the generation.
+            attention_mask (*tf.Tensor* of shape *(batch_size, sequence_length)*, *optional*):
+                Mask to avoid performing attention on padding token indices. Mask values selected in `[0, 1]`:
```
Comment (amyeroberts) [inline src/transformers/models/blip/modeling_tf_blip.py:1252]: nit
```suggestion
            pixel_values (`tf.Tensor` of shape `(batch_size, image_width, image_height)`:
                Input image to be processed
            input_ids (`tf.Tensor` of shape `(batch_size, sequence_length)`, *optional*):
                The sequence used as a prompt for the generation.
            attention_mask (`tf.Tensor` of shape `(batch_size, sequence_length)`, *optional*):
                Mask to avoid performing attention on padding token indices. Mask values selected in `[0, 1]`:
```

File: src/transformers/models/blip/modeling_tf_blip.py [hunk 2]
```diff
@@ -0,0 +1,1778 @@
...
+        >>> inputs = processor(images=image, return_tensors="tf")
+
+        >>> outputs = model.generate(**inputs)
+        >>> print(processor.decode(outputs[0], skip_special_tokens=True))
+        two cats are laying on a couch
+        ```
+        """
+
+        batch_size = pixel_values.shape[0]
+        vision_outputs = self.vision_model(
```
Comment (amyeroberts) [inline src/transformers/models/blip/modeling_tf_blip.py:1276]: nit: can be put on one line

File: src/transformers/models/blip/modeling_tf_blip.py [hunk 3]
```diff
@@ -0,0 +1,1778 @@
...
+        Parameters:
+            input_ids (*tf.Tensor* of shape *(batch_size, sequence_length)*):
+                The sequence used as a prompt for the generation.
+            pixel_values (*tf.Tensor* of shape *(batch_size, image_width, image_height)*:
+                Input image to be processed
+            attention_mask (*tf.Tensor* of shape *(batch_size, sequence_length)*, *optional*):
+                Mask to avoid performing attention on padding token indices. Mask values selected in `[0, 1]`. `1` for
+                tokens that are NOT MASKED, `0` for MASKED tokens.
+            **generate_kwargs:
+                Additional arguments passed to the *generate* function of the decoder
```
Comment (amyeroberts) [inline src/transformers/models/blip/modeling_tf_blip.py:1534]:
```suggestion
            input_ids (`tf.Tensor` of shape `(batch_size, sequence_length)`):
                The sequence used as a prompt for the generation.
            pixel_values (`tf.Tensor` of shape `(batch_size, image_width, image_height)`:
                Input image to be processed
            attention_mask (`tf.Tensor` of shape `(batch_size, sequence_length)`, *optional*):
                Mask to avoid performing attention on padding token indices. Mask values selected in `[0, 1]`. `1` for
                tokens that are NOT MASKED, `0` for MASKED tokens.
            generate_kwargs (dict, *optional*):
                Additional arguments passed to the `generate` function of the decoder
```

File: src/transformers/models/blip/modeling_tf_blip.py [hunk 4]
```diff
@@ -0,0 +1,1778 @@
...
+        attention_mask: Optional[tf.Tensor] = None,
+        **generate_kwargs,
+    ) -> tf.Tensor:
+        r"""
+        Overrides *generate* function to be able to use the model as a conditional generator
+
+        Parameters:
+            input_ids (*tf.Tensor* of shape *(batch_size, sequence_length)*):
+                The sequence used as a prompt for the generation.
+            pixel_values (*tf.Tensor* of shape *(batch_size, image_width, image_height)*:
```
Comment (amyeroberts) [inline src/transformers/models/blip/modeling_tf_blip.py:1528]: Are these dimensions right?

File: src/transformers/models/blip/modeling_tf_blip.py [hunk 5]
```diff
@@ -0,0 +1,1778 @@
...
+        >>> processor = AutoProcessor.from_pretrained("Salesforce/blip-image-captioning-base")
+
+        >>> url = "http://images.cocodataset.org/val2017/000000039769.jpg"
+        >>> image = Image.open(requests.get(url, stream=True).raw)
+
+        >>> inputs = processor(images=image, return_tensors="pt")
+
+        >>> image_features = model.get_image_features(**inputs)
+        ```"""
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
```
Comment (amyeroberts) [inline src/transformers/models/blip/modeling_tf_blip.py:1084]: Same here re `return_dict`

File: src/transformers/models/blip/modeling_tf_blip.py [hunk 6]
```diff
@@ -0,0 +1,1778 @@
...
+    """
+
+    itm_score: Optional[tf.Tensor] = None
+    loss: Optional[tf.Tensor] = None
+    image_embeds: Optional[tf.Tensor] = None
+    last_hidden_state: tf.Tensor = None
+    hidden_states: Optional[Tuple[tf.Tensor]] = None
+    vision_pooler_output: Optional[tf.Tensor] = None
+    attentions: Optional[Tuple[tf.Tensor]] = None
+    question_embeds: Optional[Tuple[tf.Tensor]] = None
```
Comment (amyeroberts) [inline src/transformers/models/blip/modeling_tf_blip.py:186]: This ordering is a bit non-standard, especially compared to the other modeling outputs in this file e.g. TFBlipForConditionalGenerationModelOutput. I know this is copied from the PyTorch model, but I wonder if it breaks certain assumptions in the codebase?

File: src/transformers/models/blip/modeling_tf_blip_text.py [hunk 7]
```diff
@@ -0,0 +1,959 @@
...
+
+
+# Adapted from https://github.com/salesforce/BLIP/blob/main/models/med.py#242
+class TFBlipTextAttention(Layer):
+    def __init__(self, config, is_cross_attention=False, **kwargs):
+        super().__init__(**kwargs)
+        self.self = TFBlipTextSelfAttention(config, is_cross_attention, name="self")
+        # "output" is a protected attribute on TF models
+        self.self_output = TFBlipTextSelfOutput(config, name="output")
+        self.pruned_heads = set()
```
Comment (amyeroberts) [inline src/transformers/models/blip/modeling_tf_blip_text.py:252]: Is this attribute needed?

File: src/transformers/models/blip/modeling_tf_blip_text.py [hunk 8]
```diff
@@ -0,0 +1,959 @@
...
+        hidden_states = self.LayerNorm(inputs=hidden_states + input_tensor)
+
+        return hidden_states
+
+
+class TFBlipTextLayer(Layer):
+    def __init__(self, config, **kwargs):
+        super().__init__(**kwargs)
+        self.config = config
+        self.seq_len_dim = 1
```
Comment (amyeroberts) [inline src/transformers/models/blip/modeling_tf_blip_text.py:322]: Is this attribute needed?

File: src/transformers/models/blip/modeling_tf_blip_text.py [hunk 9]
```diff
@@ -0,0 +1,959 @@
...
+        # The output weights are the same as the input embeddings, but there is
+        # an output-only bias for each token.
+        self.decoder = Dense(
+            config.vocab_size, kernel_initializer=get_initializer(config.initializer_range), name="decoder"
+        )
+        # TODO Make sure we can copy weights from torch despite them doing this weird shit
+        # self.bias = nn.Parameter(torch.zeros(config.vocab_size))
+
+        # Need a link between the two variables so that the bias is correctly resized with `resize_token_embeddings`
+        # self.decoder.bias = self.bias
```
Comment (amyeroberts) [inline src/transformers/models/blip/modeling_tf_blip_text.py:510]: Is this now done?

File: src/transformers/models/blip/modeling_tf_blip_text.py [hunk 10]
```diff
@@ -0,0 +1,959 @@
...
+    ) -> tf.Tensor:
+        """
+        Makes broadcastable attention and causal masks so that future and masked tokens are ignored.
+
+        Arguments:
+            attention_mask (`tf.Tensor`):
+                Mask with ones indicating tokens to attend to, zeros for tokens to ignore.
+            input_shape (`Tuple[int]`):
+                The shape of the input to the model.
+
```
Comment (amyeroberts) [inline src/transformers/models/blip/modeling_tf_blip_text.py:601]: `is_decoder` is missing here

File: src/transformers/models/blip/modeling_tf_blip_text.py [hunk 11]
```diff
@@ -0,0 +1,959 @@
...
+        # masked positions, this operation will create a tensor which is 0.0 for
+        # positions we want to attend and -10000.0 for masked positions.
+        # Since we are adding it to the raw scores before the softmax, this is
+        # effectively the same as removing these entirely.
+        extended_attention_mask = tf.cast(extended_attention_mask, self.dtype)  # fp16 compatibility
+        extended_attention_mask = (1.0 - extended_attention_mask) * -10000.0
+        return extended_attention_mask
+
+    @unpack_inputs
+    def call(
```
Comment (amyeroberts) [inline src/transformers/models/blip/modeling_tf_blip_text.py:653]: I think this might be missing the `add_start_docstring` decorator

File: src/transformers/models/blip/modeling_tf_blip_text.py [hunk 12]
```diff
@@ -0,0 +1,959 @@
...
+
+        return TFCausalLMOutputWithCrossAttentions(
+            logits=output.logits,
+            cross_attentions=output.cross_attentions,
+            hidden_states=hs,
+            attentions=attns,
+        )
+
+    @unpack_inputs
+    def call(
```
Comment (amyeroberts) [inline src/transformers/models/blip/modeling_tf_blip_text.py:845]: Same here re `add_start_docstring`

**Review by amyeroberts — COMMENTED** [review]
Nice 🔥 Thanks for adding this model!  

Mostly nits. The main comment is question whether TODO comments have been resolved.

File: src/transformers/models/blip/modeling_tf_blip.py (hunk 1 above)
Comment (gante) [reply src/transformers/models/blip/modeling_tf_blip.py:1252]: (needs update on the PT side as well)

File: src/transformers/models/blip/modeling_tf_blip.py (hunk 2 above)
Comment (gante) [reply src/transformers/models/blip/modeling_tf_blip.py:1276]: On the PT side as well

File: src/transformers/models/blip/modeling_tf_blip.py [hunk 13]
```diff
@@ -0,0 +1,1778 @@
...
+        >>> text = "How many cats are in the picture?"
+        >>> inputs = processor(images=image, text=text, return_tensors="tf")
+        >>> outputs = model.generate(**inputs)
+        >>> print(processor.decode(outputs[0], skip_special_tokens=True))
+        2
+        ```"""
+        if labels is None and decoder_input_ids is None:
+            raise ValueError(
+                "Either `decoder_input_ids` or `labels` should be passed when calling `forward` with"
+                " `BlipForQuestionAnswering`. if you are training the model make sure that `labels` is passed, if you"
```
Comment (gante) [inline src/transformers/models/blip/modeling_tf_blip.py:1456]:
```suggestion
                " `TFBlipForQuestionAnswering`. if you are training the model make sure that `labels` is passed, if you"
```

File: src/transformers/models/blip/modeling_tf_blip.py (hunk 3 above)
Comment (gante) [reply src/transformers/models/blip/modeling_tf_blip.py:1534]: corrections here should also go to the PT side

File: src/transformers/models/blip/modeling_tf_blip.py (hunk 4 above)
Comment (gante) [reply src/transformers/models/blip/modeling_tf_blip.py:1528]: It should include the number of channels. BTW, this error is present in other places in this file, as well as on the PT side

File: src/transformers/models/blip/modeling_tf_blip_text.py [hunk 14]
```diff
@@ -0,0 +1,959 @@
...
+            name="word_embeddings",
+        )
+        self.position_embeddings = Embedding(
+            config.max_position_embeddings,
+            config.hidden_size,
+            embeddings_initializer=get_initializer(config.initializer_range),
+            name="position_embeddings",
+        )
+
+        # self.LayerNorm is not snake-cased to stick with TensorFlow model variable name and be able to load
```
Comment (gante) [inline src/transformers/models/blip/modeling_tf_blip_text.py:64]:
```suggestion
        # self.LayerNorm is not snake-cased to stick with PyTorch model variable name and be able to load
```
?

File: src/transformers/models/blip/modeling_tf_blip_text.py [hunk 15]
```diff
@@ -0,0 +1,959 @@
...
+                relative_position_scores_key = tf.einsum("bhrd,lrd->bhlr", key_layer, positional_embedding)
+                attention_scores = attention_scores + relative_position_scores_query + relative_position_scores_key
+
+        attention_scores = attention_scores / math.sqrt(self.attention_head_size)
+        if attention_mask is not None:
+            # Apply the attention mask is (precomputed for all layers in BlipTextModel forward() function)
+            attention_scores = attention_scores + tf.cast(attention_mask, attention_scores.dtype)
+
+        # Normalize the attention scores to probabilities.
+        attention_probs = tf.nn.softmax(attention_scores, axis=-1)
```
Comment (gante) [inline src/transformers/models/blip/modeling_tf_blip_text.py:204]: `stable_softmax` plz, to save us from weird XLA issues

File: src/transformers/models/blip/modeling_tf_blip_text.py (hunk 7 above)
Comment (gante) [reply src/transformers/models/blip/modeling_tf_blip_text.py:252]: I don't think so, I don't think we support pruned heads in TF :D

File: src/transformers/models/blip/modeling_tf_blip_text.py (hunk 9 above)
Comment (gante) [reply src/transformers/models/blip/modeling_tf_blip_text.py:510]: See `BiasLayer` in TFBart!

**Review by gante — COMMENTED** [review]
Oof, this is a long one -- good job on getting it to the finish line! It should be close to a mergeable state.

A few general comments:
1. Missing: new modeling files in doctests;
2. The PR has some minor issues that came from the PT implementation. It would be nice to correct them as well!
3. The `training` argument is missing 😱 It needs to be added throughout the code.

File: tests/models/blip/test_modeling_tf_blip.py [hunk 16]
```diff
@@ -0,0 +1,839 @@
...
+            self.assertTrue(x is None or isinstance(x, tf.keras.layers.Layer))
+
+    def test_model(self):
+        config_and_inputs = self.model_tester.prepare_config_and_inputs()
+        self.model_tester.create_and_check_model(*config_and_inputs)
+
+    def test_training(self):
+        pass
+
+    def test_training_gradient_checkpointing(self):
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply tests/models/blip/test_modeling_tf_blip.py:184]: Nope - even if we tried it'd all get compiled and run simultaneously anyway. I'll remove this!

File: tests/models/blip/test_modeling_tf_blip.py [hunk 17]
```diff
@@ -0,0 +1,839 @@
...
+        pass
+
+    @slow
+    def test_model_from_pretrained(self):
+        for model_name in TF_BLIP_PRETRAINED_MODEL_ARCHIVE_LIST[:1]:
+            try:
+                model = TFBlipVisionModel.from_pretrained(model_name)
+            except OSError:
+                model = TFBlipVisionModel.from_pretrained(model_name, from_pt=True)
+            self.assertIsNotNone(model)
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply tests/models/blip/test_modeling_tf_blip.py:202]: Correct!

File: src/transformers/models/blip/modeling_tf_blip.py [hunk 18]
```diff
@@ -0,0 +1,1778 @@
...
+    ) -> Tuple[tf.Tensor, Optional[tf.Tensor], Optional[Tuple[tf.Tensor]]]:
+        """Input shape: Batch x Time x Channel"""
+
+        bsz, tgt_len, embed_dim = shape_list(hidden_states)
+
+        mixed_qkv = self.qkv(hidden_states)
+        mixed_qkv = tf.reshape(mixed_qkv, (bsz, tgt_len, 3, self.num_heads, self.head_dim))
+        mixed_qkv = tf.transpose(mixed_qkv, perm=(2, 0, 3, 1, 4))
+
+        query_states, key_states, value_states = (
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply src/transformers/models/blip/modeling_tf_blip.py:380]: Fixed (for both)!

File: tests/models/blip/test_modeling_tf_blip.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply tests/models/blip/test_modeling_tf_blip.py:184]: Removed!

File: tests/models/blip/test_modeling_tf_blip.py [hunk 19]
```diff
@@ -0,0 +1,839 @@
...
+            model = model_class(config)
+            self.assertIsInstance(model.get_input_embeddings(), (tf.keras.layers.Layer))
+            x = model.get_output_embeddings()
+            self.assertTrue(x is None or isinstance(x, tf.keras.layers.Layer))
+
+    def test_model(self):
+        config_and_inputs = self.model_tester.prepare_config_and_inputs()
+        self.model_tester.create_and_check_model(*config_and_inputs)
+
+    def test_training(self):
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply tests/models/blip/test_modeling_tf_blip.py]: Removed!

File: src/transformers/models/blip/modeling_tf_blip.py (hunk 2 above)
Comment (Rocketknight1) [reply src/transformers/models/blip/modeling_tf_blip.py:1265]: Fixed (for both)!

File: src/transformers/models/blip/modeling_tf_blip.py (hunk 6 above)
Comment (Rocketknight1) [reply src/transformers/models/blip/modeling_tf_blip.py:185]: I see the issue here, but I really hope we're not depending on input argument ordering - hopefully all args after the first are being called as keyword arguments!

File: src/transformers/models/blip/modeling_tf_blip.py (hunk 4 above)
Comment (Rocketknight1) [reply src/transformers/models/blip/modeling_tf_blip.py]: Updated!

File: src/transformers/models/blip/modeling_tf_blip_text.py (hunk 7 above)
Comment (Rocketknight1) [reply src/transformers/models/blip/modeling_tf_blip_text.py:255]: Removed!

File: src/transformers/models/blip/modeling_tf_blip_text.py (hunk 8 above)
Comment (Rocketknight1) [reply src/transformers/models/blip/modeling_tf_blip_text.py:324]: Removed!

File: utils/check_repo.py [hunk 20]
```diff
@@ -143,6 +143,7 @@
     "TFSegformerDecodeHead",  # Not a regular model.
     "AltRobertaModel",  # Building part of bigger (tested) model.
     "BlipTextLMHeadModel",  # No need to test it as it is tested by BlipTextVision models
+    "TFBlipTextLMHeadModel",  # No need to test it as it is tested by BlipTextVision models
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply utils/check_repo.py:146]: The authors leave this as a fruitful direction for future work

**Comment by Rocketknight1** [comment]
Got through a lot of the comments today, but I have a couple of other things to do - will try to finish them tomorrow!

**Comment by Rocketknight1** [comment]
The last remaining big issue is that some of the pt-tf equivalence tests fail when weights don't match up between models. This is caused by the cross-attention weights not being built, presumably because those layers aren't being called in the forward pass. I'm working on figuring out why and resolving that!

**Comment by Rocketknight1** [comment]
The issue seems to be that in all of our other models, cross-attention layers are only added when `config.add_cross_attention` is True, but in the case of BLIP it only checks `config.is_decoder`. As a result, the PyTorch models often initialize cross-attention layers that aren't used, which causes weight mismatch issues for us in crossloading tests, because TF only creates weights on first use.

**Review by gante — COMMENTED** [review]
Looks good!

Two high-level items from the previous review remaining:
1. Missing: new modeling files in doctests;
2. The training argument is missing 😱 It needs to be added throughout the code.

**Comment by Rocketknight1** [comment]
It's coming, don't worry! This cross-attention behaviour is just very odd and I'm trying to track it down first

File: src/transformers/models/blip/modeling_tf_blip_text.py (hunk 9 above)
Comment (Rocketknight1) [reply src/transformers/models/blip/modeling_tf_blip_text.py]: This is resolved now!

File: src/transformers/models/blip/modeling_tf_blip_text.py [hunk 21]
```diff
@@ -0,0 +1,959 @@
...
+#     https://opensource.org/licenses/BSD-3-Clause
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# TODO Check whether 'training' is propagated correctly without us manually specifying it
+# TODO Propagate name correctly
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply src/transformers/models/blip/modeling_tf_blip_text.py]: This has been implemented now

File: tests/models/blip/test_modeling_tf_blip.py [hunk 22]
```diff
@@ -0,0 +1,839 @@
...
+                expected_arg_names = (
+                    ["input_ids"] if model_class != TFBlipForConditionalGeneration else ["pixel_values"]
+                )
+                self.assertListEqual(arg_names[:1], expected_arg_names)
+
+    @unittest.skip(reason="Tested in individual model tests")
+    def test_compile_tf_model(self):
+        pass
+
+    @unittest.skip("Has some odd input names!")
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply tests/models/blip/test_modeling_tf_blip.py:708]: Done!

File: src/transformers/models/blip/modeling_tf_blip.py [hunk 23]
```diff
@@ -0,0 +1,1778 @@
...
+        inputs_embeds,
+        attention_mask: Optional[tf.Tensor] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+    ) -> Union[Tuple, TFBaseModelOutput]:
+        r"""
+        Args:
+            inputs_embeds (`tf.Tensor` of shape `(batch_size, sequence_length, hidden_size)`):
+                Optionally, instead of passing `input_ids` you can choose to directly pass an embedded representation.
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply src/transformers/models/blip/modeling_tf_blip.py:622]: Fixed in both!

File: src/transformers/models/blip/modeling_tf_blip_text.py (hunk 10 above)
Comment (Rocketknight1) [reply src/transformers/models/blip/modeling_tf_blip_text.py:609]: Added!

File: src/transformers/modeling_tf_utils.py [hunk 24]
```diff
@@ -1157,6 +1158,65 @@ def _from_config(cls, config, **kwargs):
         """
         return cls(config, **kwargs)
 
+    def invert_attention_mask(self, encoder_attention_mask: tf.Tensor) -> tf.Tensor:
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply src/transformers/modeling_tf_utils.py:1161]: There don't seem to be strong opinions here, so I'm gonna mirror the PT layout and change it later if needed!

**Comment by Rocketknight1** [comment]
cc @sgugger - I think this is ready for a final review at last!

**Review by gante — APPROVED** [review]
It has my blessing 🪄

File: src/transformers/models/blip/modeling_tf_blip.py [hunk 25]
```diff
@@ -0,0 +1,1754 @@
...
+        context_layer = tf.reshape(context_layer, new_context_layer_shape)
+
+        output = self.projection(context_layer)
+
+        outputs = (output, attention_probs) if output_attentions else (output, None)
+
+        return outputs
+
+
+# Adapted from transformers.models.clip.modeling_tf_clip.TFCLIPMLP with CLIP->Blip
```
Comment (sgugger) [inline src/transformers/models/blip/modeling_tf_blip.py:413]:
```suggestion
```
It's either Copied from or we don't care ;-)

File: src/transformers/models/blip/modeling_tf_blip_text.py [hunk 26]
```diff
@@ -0,0 +1,1015 @@
...
+        new_context_layer_shape = shape_list(context_layer)[:-2] + [self.all_head_size]
+        context_layer = tf.reshape(context_layer, new_context_layer_shape)
+
+        outputs = (context_layer, attention_probs) if output_attentions else (context_layer,)
+
+        outputs = outputs + (past_key_value,)
+        return outputs
+
+
+# Adapted from transformers.models.bert.modeling_tf_bert.TFBertSelfOutput
```
Comment (sgugger) [inline src/transformers/models/blip/modeling_tf_blip_text.py:260]:
```suggestion
```

File: src/transformers/models/blip/modeling_tf_blip_text.py [hunk 27]
```diff
@@ -0,0 +1,1015 @@
...
+            self.intermediate_act_fn = config.hidden_act
+
+    def call(self, hidden_states: tf.Tensor) -> tf.Tensor:
+        hidden_states = self.dense(inputs=hidden_states)
+        hidden_states = self.intermediate_act_fn(hidden_states)
+
+        return hidden_states
+
+
+# Adapted from transformers.models.bert.modeling_tf_bert.TFBertOutput
```
Comment (sgugger) [inline src/transformers/models/blip/modeling_tf_blip_text.py:334]:
```suggestion
```

File: tests/models/blip/test_modeling_blip.py [hunk 28]
```diff
@@ -342,6 +342,10 @@ def test_model_from_pretrained(self):
             model = BlipTextModel.from_pretrained(model_name)
             self.assertIsNotNone(model)
 
+    @unittest.skip(reason="This test class covers encoder-decoder models that the base test does not work with.")
+    def test_pt_tf_model_equivalence(self):
+        pass
```
Comment (sgugger) [inline tests/models/blip/test_modeling_blip.py:347]: Needs to be rewritten then. We cannot skip this test.

File: tests/models/blip/test_modeling_blip.py [hunk 29]
```diff
@@ -524,6 +528,10 @@ def test_model_from_pretrained(self):
             model = BlipModel.from_pretrained(model_name)
             self.assertIsNotNone(model)
 
+    @unittest.skip(reason="This test class covers encoder-decoder models that the base test does not work with.")
+    def test_pt_tf_model_equivalence(self):
```
Comment (sgugger) [inline tests/models/blip/test_modeling_blip.py:532]: Same here.

File: tests/models/blip/test_modeling_blip_text.py [hunk 30]
```diff
@@ -164,3 +164,7 @@ def test_model_from_pretrained(self):
         for model_name in BLIP_PRETRAINED_MODEL_ARCHIVE_LIST[:1]:
             model = BlipTextModel.from_pretrained(model_name)
             self.assertIsNotNone(model)
+
+    @unittest.skip(reason="This test class covers encoder-decoder models that the base test does not work with.")
+    def test_pt_tf_model_equivalence(self):
```
Comment (sgugger) [inline tests/models/blip/test_modeling_blip_text.py:169]: Same there

File: tests/models/blip/test_modeling_tf_blip.py [hunk 31]
```diff
@@ -0,0 +1,827 @@
...
+    def test_model_from_pretrained(self):
+        for model_name in TF_BLIP_PRETRAINED_MODEL_ARCHIVE_LIST[:1]:
+            try:
+                model = TFBlipTextModel.from_pretrained(model_name)
+            except OSError:
+                model = TFBlipTextModel.from_pretrained(model_name, from_pt=True)
+            self.assertIsNotNone(model)
+
+    @unittest.skip(reason="This test class covers encoder-decoder models that the base test does not work with.")
+    def test_pt_tf_model_equivalence(self):
```
Comment (sgugger) [inline tests/models/blip/test_modeling_tf_blip.py:330]: No no no no.

**Review by sgugger — COMMENTED** [review]
Thanks for your PR! It sadly cannot be merged until the pt/tf equivalence tests are all passing. There is no model tester that skips them in the code base, so let's not BLIP be the first one.

If the fact BLIP is an encoder/decoder make changes necessary to the base tests in the model tester classes. The test can be overwritten.

**Comment by Rocketknight1** [comment]
Got it, I'll figure out some way to re-enable those tests, or override them with versions that do work!

File: tests/models/blip/test_modeling_blip.py (hunk 28 above)
Comment (Rocketknight1) [reply tests/models/blip/test_modeling_blip.py:347]: Re-enabled!

File: tests/models/blip/test_modeling_blip.py (hunk 29 above)
Comment (Rocketknight1) [reply tests/models/blip/test_modeling_blip.py:532]: Re-enabled!

File: tests/models/blip/test_modeling_blip_text.py (hunk 30 above)
Comment (Rocketknight1) [reply tests/models/blip/test_modeling_blip_text.py:169]: Re-enabled!

File: tests/models/blip/test_modeling_tf_blip.py (hunk 31 above)
Comment (Rocketknight1) [reply tests/models/blip/test_modeling_tf_blip.py]: Re-enabled!

File: src/transformers/modeling_tf_utils.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply src/transformers/modeling_tf_utils.py]: Done! I didn't touch the PyTorch side yet because that's a bigger refactor that touches several models, but I can do it in another PR after this if you want.

**Comment by Rocketknight1** [comment]
@sgugger this should be ready for review with all comments addressed! The failing test is in an unrelated model

File: src/transformers/commands/pt_to_tf.py [hunk 32]
```diff
@@ -196,7 +196,7 @@ def __init__(
         self._extra_commit_description = extra_commit_description
         self._override_model_class = override_model_class
 
-    def get_inputs(self, pt_model, config):
+    def get_inputs(self, pt_model, tf_dummy_inputs, config):
```
Comment (sgugger) [inline src/transformers/commands/pt_to_tf.py:199]: The changes here seem unrelated to this PR and would be better in their own PR, no?

File: tests/models/blip/test_modeling_blip_text.py (hunk 30 above)
Comment (sgugger) [reply tests/models/blip/test_modeling_blip_text.py:169]: This one isn't either.

**Review by sgugger — COMMENTED** [review]
Still some equivalence tests missing.

**Comment by Rocketknight1** [comment]
@sgugger Sorry for the confusion - that equivalence test is present in both the `test_modeling_tf_blip` and `test_modeling_blip` file. Do we want to keep it in both?

**Comment by sgugger** [comment]
Yes we do.

File: src/transformers/commands/pt_to_tf.py (hunk 32 above)
Comment (Rocketknight1) [reply src/transformers/commands/pt_to_tf.py:199]: Fair! I added them because they were needed for the `pt-to-tf` code to port the BLIP models correctly. If you'd rather I move them to a separate PR though, that's fine!

**Review by sgugger — APPROVED** [review]
Failing tests are unrelated.

**Review by gante — APPROVED** [review]
(`pt-to-tf` changes LGTM 👍 )

**Comment by Rocketknight1** [comment]
Going to leave the `pt-to-tf` changes in this PR rather than making a separate one, since they're needed for proper BLIP conversion!
